### PR TITLE
Test that Autoscale times out building servers

### LINKED ISSUE #1319: CATC10: Server Timeout on build
Server Timeout on build
	- Validate that the expected group state is reached when servers timeout building
		- (ex. use mimic to Timeout 1/5 servers)

Prereq: Need to be able to configure Mimic to transition a subset of servers to build (building indefinitely)

### PR DESCRIPTION
**Pull request by cyli**
Fixes #1319

### CODE REVIEW

**Comment by cyli** [comment]
This is skipped because this would time out automated testing, but when I manually set autoscale to time out servers after 30 seconds, these tests pass.  (This PR implements 1 test, but it is copied over to `ConvergenceTestsWith1CLB` also).

**Comment by cyli** [comment]
The build time out config option has been merged - waiting on the ci changes before running jenkins tests.

File: otter/integration/tests/test_convergence.py [hunk 1]
```diff
@@ -853,6 +857,71 @@ def test_convergence_fixes_errored_building_servers(self):
         d.addCallback(lambda _: group)
         return d
 
+    @tag("CATC-010")
+    @inlineCallbacks
```
Comment (manishtomar) [inline otter/integration/tests/test_convergence.py]: +1 for using this. I think all tests can benefit from this.

File: otter/integration/tests/test_convergence.py [hunk 2]
```diff
@@ -853,6 +857,71 @@ def test_convergence_fixes_errored_building_servers(self):
...
+
+        1. Mimic should cause a single server to remain in building too
+           long.
+        2. Create group with 2 servers.  (One of them will time out building.)
+        3. Check with Nova that 2 servers are built - wait for one to be
+           active.  The other is the one that should remain in build.
+        4. Wait for autoscale to show 2 servers being active.
+        5. Check with Nova to ensure that there are only 2 active servers on
+           the account.  The one that was building forever should be deleted.
+        """
```
Comment (manishtomar) [inline otter/integration/tests/test_convergence.py]: +1. Nice docs :)

File: otter/integration/tests/test_convergence.py [hunk 3]
```diff
@@ -853,6 +857,71 @@ def test_convergence_fixes_errored_building_servers(self):
...
+
+        yield MimicNova(pool=self.helper.pool).sequenced_behaviors(
+            self.rcs,
+            criteria=[{"server_name": "build-timeout.*"}],
+            behaviors=[
+                {"name": "build",
+                 "parameters": {"duration": otter_build_timeout * 2}},
+                {"name": "default"}
+            ])
+        yield group.start(self.rcs, self)
```
Comment (manishtomar) [inline otter/integration/tests/test_convergence.py]: Non-blocking question irrelevant to this PR: Is it possible to return group directly from `self.helper.create_group`?

File: otter/integration/tests/test_convergence.py [hunk 4]
```diff
@@ -853,6 +857,71 @@ def test_convergence_fixes_errored_building_servers(self):
...
+
+        initial_servers = yield wait_for_servers(
+            self.rcs, pool=self.helper.pool, group=group,
+            timeout=otter_build_timeout,
+            matcher=MatchesSetwise(
+                ContainsDict({'status': Equals('ACTIVE')}),
+                ContainsDict({'status': Equals('BUILD')}),
+            ))
+        building_server_id = [s['id'] for s in initial_servers
+                              if s['status'] == 'BUILD'][0]
```
Comment (manishtomar) [inline otter/integration/tests/test_convergence.py]: Non-blocking nitpick: I've found using `next` handy: `next(s['id'] for s in initial_servers if s['status'] == 'BUILD')`

File: otter/integration/tests/test_convergence.py [hunk 5]
```diff
@@ -853,6 +857,71 @@ def test_convergence_fixes_errored_building_servers(self):
...
+                ExcludesServers([building_server_id])),
+            timeout=600)
+
+        yield wait_for_servers(
+            self.rcs, pool=self.helper.pool, group=group,
+            timeout=600,
+            matcher=MatchesAll(
+                AllMatch(ContainsDict({'status': Equals('ACTIVE'),
+                                       'id': NotEquals(building_server_id)})),
+                HasLength(2)
```
Comment (manishtomar) [inline otter/integration/tests/test_convergence.py]: Love the way you are using matchers :)

**Comment by manishtomar** [comment]
LGreatTM modulo non-blocking nitpicks. Very well written test :)

File: otter/integration/tests/test_convergence.py (hunk 4 above)
Comment (cyli) [inline otter/integration/tests/test_convergence.py]: Oh nice, thanks!

File: otter/integration/tests/test_convergence.py (hunk 3 above)
Comment (cyli) [inline otter/integration/tests/test_convergence.py]: Do you mean start the group as soon as it's created?  (`self.helper.create_group` does return the group model, but does not actually send a create request to autoscale)

Comment (manishtomar) [inline otter/integration/tests/test_convergence.py]: Yes. I am sorry I do not understand the reason for separating it into creation of model and then starting. Please note that this is irrelevant to this PR and we can talk somewhere else :)

Comment (cyli) [inline otter/integration/tests/test_convergence.py]: Yeah that is probably a separate issue.  :)  Thanks!

**Comment by cyli** [comment]
jenkins please run integration tests

**Comment by cyli** [comment]
jenkins please run integration tests

**Comment by cyli** [comment]
jenkins please run integration tests

**Comment by cyli** [comment]
Thanks for the review @manishtomar!

**Pull request merged**